### DISCUSSION TITLE
float into variables

### DISCUSSION THREAD

**Discussion by chicory-ru** [micropython/micropython#14364] [General]
count = 0
for i in range(60):
    count += 0.1
    print(count)
   
displays:
0.1
0.2
0.3
0.4
0.5
0.6
0.7
0.8000001
0.9000001
1.0
1.1
1.2
1.3
1.4
1.5
1.6
1.7
1.8
1.9
2.0
2.1
2.2
2.3
2.4
2.5
2.6
2.7
2.799999
2.899999
2.999999
3.099999
3.199999
3.299999
3.399999
3.499999
3.599999
3.699999
3.799999
3.899998
3.999998
4.099998
4.199998
4.299998
4.399998
4.499998
4.599998
4.699998
4.799998
4.899998
4.999998
5.099998
5.199997
5.299997
5.399997
5.499997
5.599997
5.699997
5.799997
5.899997
5.999997
>>> print(2.7+0.1)
2.8
>>> 

It seems that the error is accumulated in the variable. I tried it on pyboard and esp32c3, on micropython 1.22.2 and 1.23.0. How can I live with this now?

**Comment by GitHubsSilverBullet** [comment]
> How can I live with this now?

Like everybody else around the world. It's NOT a Micropython problem, it's the way floats are stored in memory.
Check IEEE754 for more information. By the way, Micropython uses regular 32bit floats (as far as I know)

PS. Try to increment by 1/16 or 1/32 or …  and be prepared to be surprised. 
Actually any increment based on the »one over power of two«.

PPS. And here's some code to visualize what's going on internally: 
```python
#!micropython

from struct import pack,unpack

def show_value(value):
    bits = unpack('I', pack('f', value))[0]
    sign = bits >> 31
    exponent = (bits>>23) & (2**8-1)
    mantisse = bits & (2**23-1)
    print(f'{value:33.30f}   S:{sign:1b} M:{mantisse:023b} E:{exponent:08b}')

def test(divider=10):
    incr = 1
    last_value, this_value = None, 1
    print('  --------decimal number---------   -----IEEE754 bitwise representation-----')
    while last_value != this_value:
        show_value(this_value)
        last_value = this_value
        incr /= divider
        this_value += incr

test(10)
print()
test(2)
```
And the result:
```
  --------decimal number---------   -----IEEE754 bitwise representation-----
  1.00000000000000000000000000000   S:0 M:00000000000000000000000 E:01111111
  1.10000002384185791015625000000   S:0 M:00011001100110011001101 E:01111111
  1.11000001430511474609375000000   S:0 M:00011100001010001111011 E:01111111
  1.11100006103515625000000000000   S:0 M:00011100011010101000000 E:01111111
  1.11110007762908935546875000000   S:0 M:00011100011100010000111 E:01111111
  1.11111009120941162109375000000   S:0 M:00011100011100011011011 E:01111111
  1.11111104488372802734375000000   S:0 M:00011100011100011100011 E:01111111
  1.11111116409301757812500000000   S:0 M:00011100011100011100100 E:01111111

  --------decimal number---------   -----IEEE754 bitwise representation-----
  1.00000000000000000000000000000   S:0 M:00000000000000000000000 E:01111111
  1.50000000000000000000000000000   S:0 M:10000000000000000000000 E:01111111
  1.75000000000000000000000000000   S:0 M:11000000000000000000000 E:01111111
  1.87500000000000000000000000000   S:0 M:11100000000000000000000 E:01111111
  1.93750000000000000000000000000   S:0 M:11110000000000000000000 E:01111111
  1.96875000000000000000000000000   S:0 M:11111000000000000000000 E:01111111
  1.98437500000000000000000000000   S:0 M:11111100000000000000000 E:01111111
  1.99218750000000000000000000000   S:0 M:11111110000000000000000 E:01111111
  1.99609375000000000000000000000   S:0 M:11111111000000000000000 E:01111111
  1.99804687500000000000000000000   S:0 M:11111111100000000000000 E:01111111
  1.99902343750000000000000000000   S:0 M:11111111110000000000000 E:01111111
  1.99951171875000000000000000000   S:0 M:11111111111000000000000 E:01111111
  1.99975585937500000000000000000   S:0 M:11111111111100000000000 E:01111111
  1.99987792968750000000000000000   S:0 M:11111111111110000000000 E:01111111
  1.99993896484375000000000000000   S:0 M:11111111111111000000000 E:01111111
  1.99996948242187500000000000000   S:0 M:11111111111111100000000 E:01111111
  1.99998474121093750000000000000   S:0 M:11111111111111110000000 E:01111111
  1.99999237060546875000000000000   S:0 M:11111111111111111000000 E:01111111
  1.99999618530273437500000000000   S:0 M:11111111111111111100000 E:01111111
  1.99999809265136718750000000000   S:0 M:11111111111111111110000 E:01111111
  1.99999904632568359375000000000   S:0 M:11111111111111111111000 E:01111111
  1.99999952316284179687500000000   S:0 M:11111111111111111111100 E:01111111
  1.99999980926513671875000000000   S:0 M:11111111111111111111110 E:01111111
  1.99999990463256835937500000000   S:0 M:11111111111111111111111 E:01111111
  2.00000000000000000000000000000   S:0 M:00000000000000000000000 E:10000000
```

**Comment by sosi-deadeye** [comment]
Increment an integer by one and to get the result, divide it by 10.

```python
count=0
for i in range(60):
    count += 1
    print(count / 10)
```

**Reply by GitHubsSilverBullet** [reply]
You are just hiding the underlying problem:
```python
count=0
for i in range(60):
    count += 1
    print(f'{count/10:33.30f}')
```
Result:
```
  0.10000000000000000000000000000
  0.20000000000000000000000000000
  0.30000000000000000000000000000
  0.40000000000000000000000000000
  0.50000000000000000000000000000
  0.60000000000000000000000000000
  0.70000000000000000000000000000
  0.80000000000000000000000000000
  0.90000000000000000000000000000
  1.00000000000000000000000000000
  1.10000002384185791015625000000
  1.20000004768371582031250000000
  1.29999995231628417968750000000
  1.39999998092651367187500000000
  1.50000000000000000000000000000
  1.60000000000000000000000000000
  1.70000004768371582031250000000
  1.79999995231628417968750000000
  1.90000000000000000000000000000
  2.00000000000000000000000000000
  2.09999990463256835937500000000
  2.20000004768371582031250000000
  2.29999995231628417968750000000
  2.40000009536743164062500000000
  2.50000000000000000000000000000
  2.59999990463256835937500000000
  2.70000004768371582031250000000
  2.79999995231628417968750000000
  2.90000009536743164062500000000
  3.00000000000000000000000000000
  3.09999990463256835937500000000
  3.20000004768371582031250000000
  3.29999995231628417968750000000
  3.40000009536743164062500000000
  3.50000000000000000000000000000
  3.59999990463256835937500000000
  3.70000004768371582031250000000
  3.79999995231628417968750000000
  3.90000009536743164062500000000
  4.00000000000000000000000000000
  4.09999990463256835937500000000
  4.19999980926513671875000000000
  4.30000019073486328125000000000
  4.40000009536743164062500000000
  4.50000000000000000000000000000
  4.59999990463256835937500000000
  4.69999980926513671875000000000
  4.80000019073486328125000000000
  4.90000009536743164062500000000
  5.00000000000000000000000000000
  5.09999990463256835937500000000
  5.19999980926513671875000000000
  5.30000019073486328125000000000
  5.40000009536743164062500000000
  5.50000000000000000000000000000
  5.59999990463256835937500000000
  5.69999980926513671875000000000
  5.80000019073486328125000000000
  5.90000009536743164062500000000
  6.00000000000000000000000000000

```

**Reply by trentp-igor** [reply]
There are really multiple problems in the OP's code, at least if you wanted it to be exact, or are interested in the numerical methods used.

The most obvious is that negative powers of ten can not be expressed in terms of negative powers of two, which is to say not all decimal values have an exact binary floating point representation.  And you're quite right, dividing an integer by ten to produce a float still has this same problem.

But there are other problems.

The value 0.1 is also inexact, and repeatedly adding an inexact value accumulates error.  By repeatedly adding an integer, this accumulation of error is avoided.  This is a general pitfall of floating-point math.  One should avoid repeatedly adding/subtracting values when it can accumulate error.  A way to avoid this is to repeatedly add/subtract integers instead.  Convert to floating point, but don't accumulate in floating-point.

Another flaw, which would become apparent as the `range()` is increased, is adding numbers of different magnitude.  This does not work well in floating-point math.  If the range was large enough, eventually `count + 0.1 == count` would be true and `count` will get no larger no matter how many times 0.1 is added to it. Even if we added `1` each time, which can be exactly represented as a float, this would happen.

**Comment by chicory-ru** [comment]
Thank you, I understand. Isn't it possible to make a data type that would store the integer part and the fractional part separately in two int variables?

**Reply by GitHubsSilverBullet** [reply]
Yes, that's called fixed-point arithmetic. But you have to remember, the number of available bits somehow determines how precise it's going to be.
If you use two 32bit integers, then there are only roughly 4billion possible values for the integer and also for the fractional part. 

Another way of thinking about it, is: Take a piece of checkered paper
You can divide 1 by 2 -> 0.5 and you only need 2 boxes.
Same goes for 1 divided by 5, or 1 div 10,
Do the same for 1 div 8 -> 0.125 and you need 4 boxes.
Now try 1 div 7 ->  7 boxes endless repeating (0.142857 142857 142857 … )
Finally 1 div 3, 1 div 6, 1 div 9 and all the paper on this planet won't allow you to write the resulting value precisely.

The same problem happens with binary values.

**Reply by ricksorensen** [reply]
See:
#9821 

For alternative number model approaches ...

**Reply by trentp-igor** [reply]
You certainly could do that.  It's part of the CPython standard library, https://docs.python.org/3/library/fractions.html

```
from fractions import Fraction
count = Fraction(0)
for i in range(60): 
    count += Fraction(1,10) 
    print(float(count)) 

0.1
0.2
0.3
0.4
0.5
0.6
0.7
0.8
0.9
1.0
1.1
1.2
1.3
1.4
1.5
1.6
1.7
1.8
1.9
2.0
2.1
…
```

But mpy doesn't have that module.

**Reply by GitHubsSilverBullet** [reply]
That would work as long as one stays within ℚ.
Unfortunately ℝ is so much bigger than ℚ.  😉

**Comment by peterhinch** [comment]
> By the way, Micropython uses regular 32bit floats (as far as I know)

On platforms where there is hardware support (e.g. Pyboard D SF6W, Unix) it uses 64 bit. On other platforms software 64 bit support can be enabled with a build option.

**Comment by chicory-ru** [comment]
Interesting. Why doesn't count += 0.5 break anything?
0.500000000000000000
1.000000000000000000
1.500000000000000000
2.000000000000000000
2.500000000000000000
3.000000000000000000
3.500000000000000000

**Reply by robert-hh** [reply]
Because 0.5 can be exactly represented as binary number. Like 0.25, 0.125, .... But not 0.1

**Reply by GitHubsSilverBullet** [reply]
Any fraction which can be built with a combination of 1/2**n values are OK.
So, for instance  1/2+1/8-1/32-1/4096 will be represented exactly.
(as long as the mantissa doesn't get greater that the number of bits, for floats 23bits) 
```python
#!micropython
# -*- coding: utf-8 -*-
result = 1 + 1 / 2**22
print(f'{result:35.30f}')
result = 1 + 1 / 2**23
print(f'{result:35.30f}')
result = 1 + 1 / 2**24
print(f'{result:35.30f}')
```
Result:
```
    1.00000023841857910156250000000
    1.00000011920928955078125000000
    1.00000000000000000000000000000
```

**Comment by sosi-deadeye** [comment]
> You are just hiding the underlying problem:

There is no inaccuracy as a result of addition because they are `int`.
But the division results in a `float` and floats are by definition inaccurate.
`round()` or f-strings should be used.

- do not compare floats of equality
- if you need to compare floats of equality, then use `math.isclose()`
- comparison is ok: `<=` `<` `>` `>=`
- use round or f-strings to represent the floats

Correction of code:
```python
count=0
for i in range(60):
    count += 1
    value = count / 10
    print(f"{value:.1f}")
```

**Reply by GitHubsSilverBullet** [reply]
> floats are by definition inaccurate.

That's so fundamentally untrue that even the opposite wouldn't be true. ;-)

Let's say that floats are a problem for those who do not understand them and their limitations.
On the other hand, that's also true for all other numerical representations on a computer system.
And it's true for floats, doubles, fixedpoint, arbitrary precision, and what else is out there.

**Reply by peterhinch** [reply]
The process of mapping floats onto a fixed size integer variable is by definition inaccurate.

I rather think this even applies if the integer has infinite size - floats being aleph 1 and integers aleph 0. I may be wrong, my memory of this stuff is hazy.

**Reply by GitHubsSilverBullet** [reply]
Peter, please map a float value of – hmmm, let's say – 3.0 to an integer and let us know about the expected and realized precision.
That goes for roughly 33 million other floats as well. ;-)
There is one number in those 33 millions where information is lost,
and that's the float number -0 (as opposed to +0). 
Mapping a float -0 to an integer looses the sign.
Just saying…

Oh, one more thing, those set of roughly 33 million number which are an absolute precise representation, those sets repeat 255 times. With intervals between two consecutive numbers of two,four,eight,… or moving the other direction 1/2,1/4,1/8,…

**Reply by peterhinch** [reply]
With a sufficiently large number of bits you can create a mapping whereby many floats have precise representations. You might even create a mapping where transcendentals such as `e` and π map onto finite bit patterns. But the general problem of mapping all floats onto integers is provably impossible, even with an infinite number of bits.

**Comment by GitHubsSilverBullet** [comment]
> ```python
>  print(f"{value:.1f}")
> ```

Again, you are not displaying the true contents of ›value‹
What  you are doing is deliberately hiding the problematic portion.
Which will become visible by using a proper format string, for instance ('value:.10f')

**Comment by bixb922** [comment]
There is a whole field of mathematics to deal with this problem. See https://en.m.wikipedia.org/wiki/Numerical_analysis, generation and propagation of errors.

Fortunately there are packages such as ulab (numpy for big python) that help to minimize the effects of the floating point error.

**Comment by sosi-deadeye** [comment]
> Again, you are not displaying the true contents of ›value‹


We don't care!
We don't want to see the real value!

I hate these dogmatic discussions.

**Reply by GitHubsSilverBullet** [reply]
> We don't care!
> We don't want to see the real value!

Wow. Clueless and loud at the same time. What a combination.

**Reply by robert-hh** [reply]
@GitHubsSilverBullet Please stay polite.

**Reply by GitHubsSilverBullet** [reply]
@robert-hh Take a look at the previous post. Reconsider.
